### ISSUE TITLE
Usage of Chamfer Loss

### ISSUE THREAD

**Issue by kongsgard** [ThibaultGROUEIX/3D-CODED#8]
Hi, and thanks for sharing a PyTorch implementation of Chamfer loss.
However, I find the returned value from calling `distChamfer(points1, points2)` to be surprisingly low when both `points1` and `points2` are two arbitrary batches of point clouds with dimension [B,N,3] where B is the batch size and N is the number of points.

I found sklearn and numpy implementations of Chamfer Distance in this stackoverflow-response (https://stackoverflow.com/questions/47060685/chamfer-distance-between-two-point-clouds-in-tensorflow).
With points1 and points2 having dimensions [32,1024,3] I typically get a loss of ~0.07 with the PyTorch implementation using
```
dist1, dist2 = distChamfer(points1, points2)
loss = (torch.mean(dist1)) + (torch.mean(dist2))
```
while both the sklearn and numpy implementations linked to above give a loss of ~0.3.

Could you point me in the right direction in how to correctly use the provided Chamfer loss implementation?
Could it potentially be a bug?

**Comment by kongsgard** [comment]
I quickly put together a script for comparison:
```
import torch
from graphs.losses.dist_chamfer import ChamferDist

cuda0 = torch.device('cuda:0')
x1 = torch.ones([1,2,3], device=cuda0)
x2 = torch.zeros([1,4,3], device=cuda0)

distChamfer = ChamferDist()
dist1, dist2 = distChamfer(x1, x2)
dist_total = torch.mean(dist1) + torch.mean(dist2)
print("PyTorch Chamfer Loss:", dist_total.item())


import numpy as np
from sklearn.neighbors import KDTree

def chamfer_distance_sklearn(array1,array2):
    batch_size, num_point = array1.shape[:2]
    dist = 0
    for i in range(batch_size):
        tree1 = KDTree(array1[i], leaf_size=num_point+1)
        tree2 = KDTree(array2[i], leaf_size=num_point+1)
        distances1, _ = tree1.query(array2[i])
        distances2, _ = tree2.query(array1[i])
        av_dist1 = np.mean(distances1)
        av_dist2 = np.mean(distances2)
        dist = dist + (av_dist1+av_dist2)/batch_size
    return dist

x1_np = np.ones([1,2,3])
x2_np = np.zeros([1,4,3])
print("sklearn Chamfer Loss:", chamfer_distance_sklearn(x1_np, x2_np))
```

This returns:
PyTorch Chamfer Loss: 6.0
sklearn Chamfer Loss: 3.46

Manual calculation validates the sklearn output.
It seems like it's simply a square root that's missing from the PyTorch implementation.
By calculating the distance with `dist_total = torch.mean(torch.sqrt(dist1)) + torch.mean(torch.sqrt(dist2))` we get the same output.

**Comment by ThibaultGROUEIX** [comment]
Hi @kongsgard ,
Yes, your analysis is correct. It is actually intentional, It corresponds to how we define it in the paper :https://arxiv.org/pdf/1806.05228.pdf
Best regards,
Thibault

**Comment by kongsgard** [comment]
I see. Thank you for your response!
